docs/design/proposals/encryption-with-vault-tokens.md

```diff
@@ -0,0 +1,125 @@
+# Multi-tenancy with Vault Tokens
+
+## Current Feature: Vault access with single token
+
+The current feature to [support Hashicorp Vault as a KMS is
+documented](./encrypted-pvc.md):
+
+- tenants can create PVCs that are encrypted with a unique key
+- the key to encrypt/decrypt the PVC is stored in Hashicorp Vault
+- there is a single token (Kubernetes ServiceAccount) to access Vault
+- enabling is done in the StorageClass, pointing to a KMS configuration section
+  in a JSON configuration file (Kubernetes ConfigMap)
+
+## High-Level Requirements
+
+The new feature should support multi-tenancy, so that each Tenant can use their
+own Vault Token for PVC encryption:
+
+- each tenant can manage their Vault token (and PVC keys)
+- to get the key from Vault, each tenant should provide their personal token
+- Ceph-CSI should use the token from the tenant to store the unique key for the
+  PVC
+- a Tenant is the owner of a Kubernetes Namespace
```

Comment:
yes, I was considering to add a referral option in the configuration file, such as 
```json
"tenants": {
    "my-app": {
        "vaultToken": "the secret value"
    },
    "companion-app": {
        "owner": "my-app"
    }
}
```
That would need to be configured by the storage admin, and may not be re-configurable by tenants themselves (otherwise any tenant would be able to decrypt PVs from other tenants).

I'd see this as a possible extension, and something that can be added at a later point in time.

Reply:
Agree, in the current scheme we can always ask the user to add a per-namespace token.